File: bindings/java/src/main/java/com/strling/simply/Pattern.java
package com.strling.simply;

import com.strling.jvm.NativeClient;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;

/** Immutable recipe that records canonical Simply operations on demand. */
public final class Pattern {
    interface Recipe {
        SimplyBuilder.Value record(Context context);
    }

    static final class Context {
        final SimplyBuilder builder;
        private int sequence;

        Context(SimplyBuilder builder) {
            this.builder = builder;
        }

        String next(String prefix) {
            sequence += 1;
            return prefix + "-" + sequence;
        }
    }

    private final Recipe recipe;
    private final Pattern repeatedValue;
    private final Integer repeatedMinimum;
    private final Integer repeatedMaximum;
    private final String repeatedMode;

    Pattern(Recipe recipe) {
        this(recipe, null, null, null, null);
    }

    private Pattern(Recipe recipe, Pattern repeatedValue, Integer repeatedMinimum,
            Integer repeatedMaximum, String repeatedMode) {
        this.recipe = Objects.requireNonNull(recipe, "recipe");
        this.repeatedValue = repeatedValue;
        this.repeatedMinimum = repeatedMinimum;
        this.repeatedMaximum = repeatedMaximum;
        this.repeatedMode = repeatedMode;
    }

    SimplyBuilder.Value record(Context context) {
        return recipe.record(context);
    }

    public Pattern repeat(int minimum, Integer maximum) {
        if (minimum < 0 || (maximum != null && maximum.intValue() < minimum)) {
            throw new STRlingError("repetition bounds must be non-negative and ordered");
        }
        Pattern source = this;
        Recipe next = context -> context.builder.repeat(
                context.next("repeat"), source.record(context), minimum, maximum, "greedy");
        return new Pattern(next, source, Integer.valueOf(minimum), maximum, "greedy");
    }

    public Pattern repeat(int count) {
        return repeat(count, Integer.valueOf(count));
    }

    public Pattern may() {
        return repeat(0, Integer.valueOf(1));
    }

    public Pattern lazy() {
        if (repeatedValue == null) {
            throw new STRlingError("lazy requires a repeated pattern");
        }
        Recipe next = context -> context.builder.repeat(
                context.next("repeat"), repeatedValue.record(context),
                repeatedMinimum.intValue(), repeatedMaximum, "lazy");
        return new Pattern(next, repeatedValue, repeatedMinimum, repeatedMaximum, "lazy");
    }

    public Pattern asCapture() {
        Pattern source = this;
        return new Pattern(context -> context.builder.capture(
                context.next("capture"), context.next("capture-key"),
                source.record(context), null));
    }

    public Pattern asGroup(String name) {
        if (name == null || name.isEmpty()) {
            throw new STRlingError("capture group name cannot be empty");
        }
        Pattern source = this;
        return new Pattern(context -> context.builder.capture(
                context.next("capture"), name, source.record(context), name));
    }

    public Map<String, Object> buildRequest(Map<String, ?> compileProjection) {
        return buildRequest(compileProjection, "java-simply", null);
    }

    public Map<String, Object> buildRequest(Map<String, ?> compileProjection,
            String identityNamespace, Map<String, ?> semanticOptions) {
        SimplyBuilder builder = semanticOptions == null
                ? new SimplyBuilder(identityNamespace)
                : new SimplyBuilder(identityNamespace, "1.0-draft.1", semanticOptions);
        Context context = new Context(builder);
        return builder.buildRequest(record(context), compileProjection);
    }

    public Object compile(NativeClient client, Map<String, ?> compileProjection,
            Map<String, ?> targetProfile) {
        return client.simplyCompile(buildRequest(compileProjection), targetProfile);
    }

    public Object compile(NativeClient client, Map<String, ?> compileProjection) {
        return compile(client, compileProjection, null);
    }

    public Object exec(String ignoredText) {
        throw new STRlingError(
                "Pattern.exec was retired: canonical adapters do not simulate runtime regex execution");
    }

    @Override
    public String toString() {
        throw new STRlingError(
                "implicit regex rendering was retired: compile through the canonical adapter");
    }

    static Pattern stdlibHelper(String helperId, Map<String, ?> parameters) {
        return new Pattern(context -> context.builder.stdlibHelper(
                context.next("stdlib"), helperId, parameters));
    }
}

File: bindings/csharp/src/STRling/Canonical/Simply/Pattern.cs
using System.Text.Json;
using Strling.Native;

namespace Strling.Simply;

/// <summary>Immutable recipe that records canonical Simply operations on demand.</summary>
public sealed class Pattern
{
    internal delegate SimplyBuilder.Value Recipe(Context context);
    internal sealed class Context(SimplyBuilder builder)
    {
        private int sequence;
        internal SimplyBuilder Builder { get; } = builder;
        internal string Next(string prefix) => $"{prefix}-{++sequence}";
    }

    private readonly Recipe recipe;
    private readonly Pattern? repeatedValue;
    private readonly int? repeatedMinimum;
    private readonly int? repeatedMaximum;

    internal Pattern(Recipe recipe, Pattern? repeatedValue = null, int? repeatedMinimum = null, int? repeatedMaximum = null)
    {
        this.recipe = recipe ?? throw new ArgumentNullException(nameof(recipe));
        this.repeatedValue = repeatedValue;
        this.repeatedMinimum = repeatedMinimum;
        this.repeatedMaximum = repeatedMaximum;
    }

    internal SimplyBuilder.Value Record(Context context) => recipe(context);

    public Pattern Repeat(int minimum, int? maximum = null)
    {
        if (minimum < 0 || (maximum.HasValue && maximum.Value < minimum))
            throw new STRlingError("repetition bounds must be non-negative and ordered");
        var source = this;
        return new Pattern(
            context => context.Builder.Repeat(context.Next("repeat"), source.Record(context), minimum, maximum, "greedy"),
            source,
            minimum,
            maximum);
    }

    public Pattern Repeat(int count) => Repeat(count, count);
    public Pattern May() => Repeat(0, 1);
    public Pattern Lazy()
    {
        if (repeatedValue is null || !repeatedMinimum.HasValue) throw new STRlingError("Lazy requires a repeated pattern");
        return new Pattern(context => context.Builder.Repeat(
            context.Next("repeat"), repeatedValue.Record(context), repeatedMinimum.Value, repeatedMaximum, "lazy"));
    }

    public Pattern Capture()
    {
        var source = this;
        return new Pattern(context => context.Builder.Capture(
            context.Next("capture"), context.Next("capture-key"), source.Record(context)));
    }

    public Pattern AsGroup(string name)
    {
        if (string.IsNullOrEmpty(name)) throw new STRlingError("capture group name cannot be empty");
        var source = this;
        return new Pattern(context => context.Builder.Capture(context.Next("capture"), name, source.Record(context), name));
    }

    public Pattern Then(params Pattern[] parts) => S.Merge([this, .. parts]);

    public JsonElement BuildRequest(
        JsonElement compileProjection,
        string identityNamespace = "csharp-simply",
        IReadOnlyDictionary<string, object?>? semanticOptions = null)
    {
        var builder = new SimplyBuilder(identityNamespace, "1.0-draft.1", semanticOptions);
        var context = new Context(builder);
        return builder.BuildRequest(Record(context), compileProjection);
    }

    public JsonElement Compile(NativeClient client, JsonElement compileProjection, JsonElement? targetProfile = null) =>
        client.SimplyCompile(BuildRequest(compileProjection), targetProfile);

    public object Exec(string ignoredText) => throw new STRlingError(
        "Pattern.Exec was retired: canonical adapters do not simulate runtime regex execution");

    public override string ToString() => throw new STRlingError(
        "implicit regex rendering was retired: compile through the canonical adapter");

    internal static Pattern StdlibHelper(string helperId, IReadOnlyDictionary<string, object?> parameters) =>
        new(context => context.Builder.StdlibHelper(context.Next("stdlib"), helperId, parameters));
}
